### ISSUE TITLE
HE-AAC: strange duraction, same ms duration in ffmpeg with original .flac

### ISSUE THREAD

**Issue by ValZapod** [MediaArea/MediaInfoLib#1570]
So this will require this very recent (2 weeks ago) commit in ffmpeg that finally fixed two times longer than needed removal of priming (media time in editlist) in HE-AAC (and v2, but not simple AAC LC) in native decoder of ffmpeg. 
https://github.com/FFmpeg/FFmpeg/commit/5114ce1e2a4c71ddf4971ad3cf9bd43ae16571c3

The duration of this file with removed priming and remainder is same as flac of this song: 3 m 58 s 800 ms. In fact mediainfo is completly right that remainder is 19 ms (though the name Duration_LastFrame is unfortunate), and in fact the size of wav decoded by ffmpeg with the commit above is 3 m 58 s 819 ms. But due to Mediainfo's duration being wrong (03:58:915), even if you subtract 19 ms you will not get to 800 ms.

Samples will be added in the next comment to avoid copyright issues with Tidal (will delete that comment when you will get the files).

**Comment by ValZapod** [comment]
https://user-images.githubusercontent.com/31514790/183337358-4c4266bc-02e4-4f81-8332-717ba6e6c186.mp4

FLAC is here: https://www.upload.ee/files/14397313/02_-_Abba_-_Don_t_Shut_Me_Down.flac.html

**Comment by JeromeMartinez** [comment]
We currently don't handle well edit lists not starting at 0 (start offset is ignored), we need to rework this code... :(

**Comment by ValZapod** [comment]
Can you tell me how to you get duration? Because this file has some quite strange 

mvhd 238800
tkhd 238800
mdhd 238914.

Yet after -c copy the file using ffmeg I do get
238915 for two first values.

**Comment by JeromeMartinez** [comment]
> Can you tell me how to you get duration?

tkhd duration is 238915 and mvhd time scale of 1000, and edit list is ignored (it should not), so 03:58:915.

> tkhd 238800

tkhd duration is 238915 in the mp4 file you provided.

**Comment by ValZapod** [comment]
> tkhd duration is 238915 in the mp4 file you provided.

Oh, it is the opposite! You get after -c copy
mvhd 238800
tkhd 238800
mdhd 238914

which is now correct since mvhd and tkhd present the post-editlist duration, while mdhd should have the pre-editlist duration. So really the file is buggy (Tidal produces such files). The problem is now there is still "source_" wrong duration coming from somewhere.

**Comment by JeromeMartinez** [comment]
> The problem is now there is still "source_" wrong duration coming from somewhere.

from the rewrapped file? I rewrapped and now I have mdhd 10536138/44100 = 238.915 (I don't know from where you get 238.914... Maybe integer truncation of 238914.7 ms?).
`Source_` info comes from stts atom (sum of times. In your file rewrapped by FFmpeg it is equal to mdhd duration).

In practice, If I understand correctly the edit list:
- media content is 238.915 s long (all frames but last one at 2048 samples, last one at 1226 samples, at least indicated at the stts level)
- playback starts at offset 0.115 s (5058 samples)
- duration is 238.800 s long

So we have 0.115 s discarded, 238.8000 s played = 238.915 s stored.

> So really the file is buggy

Good to know for avoiding me headaches when I try to fix my code. But I should handle such kind of incoherencies if it is in the wild, and show a warning that the file is buggy and try to show the real value if available (here I guess that I should peak the edit list duration).

> and in fact the size of wav decoded by ffmpeg with the commit above is 3 m 58 s 819 ms

Looks like that FFmpeg correctly discards the first 5058 samples indicated to be discarded but does not discard the last 2048 - 1226 = 822 samples ( = 0.019 s) indicated to be discarded (in both the edit list and stts).

**Comment by ValZapod** [comment]
> Looks like that FFmpeg correctly discards the first 5058 samples indicated to be discarded but does not discard the last 2048 - 1226 = 822 samples

Yes, it still does not remove remainder, so is not truely gapless. Chromium's fork of ffmpeg fixed it though, and it is a known bug. https://trac.ffmpeg.org/ticket/7828

>and stts

Where does -19 ms come from? Is it from stts?

**Comment by JeromeMartinez** [comment]
> Chromium's fork of ffmpeg fixed it though, and it is a known bug

Sad that FFmpeg team didn't merge the patch.

> Where does -19 ms come from? Is it from stts?

Yes (last frame duration minus normal duration). I need to better show when the last frame does not have the same duration as others.

**Comment by ValZapod** [comment]
>Sad that FFmpeg team didn't merge the patch.

This patch should be THE patch, can you resend it with fate fixed? https://patchwork.ffmpeg.org/project/ffmpeg/patch/20190429225027.81295-1-fumoboy007@me.com/

>Is it from stts?

What is the relationship with duration in editlist? Both should be noted?
